File: trunk/Bricksmith/UnitTests/Classes/LDraw/Commands/LDrawMetaCommand_Tests.m
//
//  LDrawMetaCommand_Tests.m
//  UnitTests
//
//  Created by Sergey Slobodenyuk on 2023-03-12.
//

#import "LDrawMetaCommand.h"

#import <XCTest/XCTest.h>
#import "LDrawColor.h"
#import "LDrawComment.h"
#import "LDrawLSynthDirective.h"
#import "LPubCommand.h"
#import "LPubRemoveGroup.h"
#import "MockArchiver.h"
#import "MockScanner.h"


// MARK: Mocks -

@interface LDrawMetaCommand ()

+ (LDrawMetaCommand *) metaCommandInstanceByMarker:(NSString *)ldrawMarker scanner:(NSScanner *)scanner;

@end


// MARK: - Tests -

@interface LDrawMetaCommand_Tests : XCTestCase

@end

@implementation LDrawMetaCommand_Tests

- (void)test_LDrawMetaCommand_ClassContainsSubclasses
{
	SEL sel = NSSelectorFromString(@"subclassNames");
	NSArray *subclasses = [LDrawMetaCommand performSelector:sel];
	XCTAssertEqual(subclasses.count, 4);
	XCTAssertTrue([subclasses containsObject:@"LDrawColor"]);
	XCTAssertTrue([subclasses containsObject:@"LDrawComment"]);
	XCTAssertTrue([subclasses containsObject:@"LDrawLSynthDirective"]);
	XCTAssertTrue([subclasses containsObject:@"LPubCommand"]);
}


- (void)test_LDrawMetaCommand_initWithLines_Success
{
	// This version uses XCTActivity to split multiple test executions with different parameters (visible only in Report Navigator).
	// To have separate lines in Test Navigator the technic with overriding testInvocations method should be used.
	// See example: https://github.com/manicmaniac/xcnew/blob/7417eb2efa2c60ccd71f3d13be8582a54d1fedd2/Tests/xcnew-tests/XCNOptionParserParameterizedTests.m
	
	NSArray *testCases = @[
		@{@"line":
			  @"  ",
		  @"result":
			  [LDrawMetaCommand class]},

		@{@"line":
			  @"",
		  @"result":
			  [LDrawMetaCommand class]},

		@{@"line":
			  @"0 some text",
		  @"result":
			  [LDrawMetaCommand class]},

		@{@"line":
			  @"0 !COLOUR name CODE 100 VALUE #F0F0F0 EDGE 400 ALPHA 0",
		  @"result":
			  [LDrawColor class]},

		@{@"line":
			  @"0 // comment",
		  @"result":
			  [LDrawComment class]},

		@{@"line":
			  @"0 !LPUB command",
		  @"result":
			  [LPubCommand class]},

		@{@"line":
			  @"0 !LPUB REMOVE GROUP \"a\"",
		  @"result":
			  [LPubRemoveGroup class]}
	];
	
	for (NSDictionary *testCase in testCases) {
		NSString *line = testCase[@"line"];
		id result = testCase[@"result"];
		NSString *activityName = [NSString stringWithFormat:@"Line: \"%@\"", line];
		
		[XCTContext runActivityNamed:activityName block:^(id<XCTActivity>  _Nonnull activity) {
			LDrawMetaCommand *command =
				[[LDrawMetaCommand alloc] initWithLines:@[line]
												inRange:NSMakeRange(0, 1)
											parentGroup:NULL];
			XCTAssertIdentical(command.class, result);
		}];
	}
}


- (void)test_LDrawMetaCommand_EncodedProperly
{
	MockArchiver *encoder = [MockArchiver new];
	LDrawMetaCommand *command = [[LDrawMetaCommand alloc] init];
	command.commandString = @"Some string";

	[command encodeWithCoder:encoder];

	XCTAssertEqual(encoder.data.count, 1);
	XCTAssertEqualObjects(encoder.data[@"commandString"], @"Some string");
}


- (void)test_LDrawMetaCommand_DecodedProperly
{
	MockArchiver *decoder = [MockArchiver new];
	decoder.data = [@{
		@"commandString" : @"Some string"
	} mutableCopy];

	LDrawMetaCommand *command = [[LDrawMetaCommand alloc] initWithCoder:decoder];
	command.commandString = @"Some string";

	XCTAssertEqualObjects(command.commandString, @"Some string");
}


- (void)test_LDrawMetaCommand_CopiedProperly
{
	LDrawMetaCommand *command = [[LDrawMetaCommand alloc] init];
	command.commandString = @"Some string";

	LDrawMetaCommand *duplicate = [command copyWithZone:nil];

	XCTAssertEqualObjects(command.commandString, duplicate.commandString);
	XCTAssertNotIdentical(command, duplicate);
}


- (void)test_LDrawMetaCommand_CopiedObjectHasNonIdenticalValues
{
	LDrawMetaCommand *command = [[LDrawMetaCommand alloc] init];
	command.commandString = @"Some string";

	LDrawMetaCommand *duplicate = [command copyWithZone:nil];
	command.commandString = @"Another string";

	XCTAssertNotEqualObjects(command.commandString, duplicate.commandString);
}


- (void)test_LDrawMetaCommand_metaCommandInstanceByMarker_ReturnsNil
{
	NSString *marker = @"string";
	id scanner = [MockScanner new];
	LDrawMetaCommand *command = [LDrawMetaCommand metaCommandInstanceByMarker:marker scanner:scanner];
	
	XCTAssertNil(command);
}

@end

File: trunk/Packages/LDrawCore/Tests/LDrawCoreTests/LDrawMetaCommandTests.swift
//
//  LDrawMetaCommandTests.swift
//  LDrawCoreTests
//
//  Which class a type 0 line becomes, and how a meta command is archived and
//  copied.
//
//  Created by Sergey Slobodenyuk on 2023-03-12.
//

import Testing
import Foundation
import LDrawCore

@Suite("Meta commands")
struct LDrawMetaCommandTests {

    @Test("The subclasses a type 0 line can become are found at run time")
    func subclassesAreFound() throws {
        let names: [String] = try #require(perform("subclassNames", on: LDrawMetaCommand.self))

        #expect(Set(names) == ["LDrawColor", "LDrawComment", "LDrawLSynthDirective", "LPubCommand"])
    }

    @Test("A line becomes the class its marker names", arguments: [
        ("  ", LDrawMetaCommand.self),
        ("", LDrawMetaCommand.self),
        ("0 some text", LDrawMetaCommand.self),
        ("0 !COLOUR name CODE 100 VALUE #F0F0F0 EDGE 400 ALPHA 0", LDrawColor.self),
        ("0 // comment", LDrawComment.self),
        ("0 !LPUB command", LPubCommand.self),
        ("0 !LPUB REMOVE GROUP \"a\"", LPubRemoveGroup.self),
    ] as [(String, AnyClass)])
    func lineBecomesItsClass(line: String, expected: AnyClass) throws {
        let command = try #require(LDrawMetaCommand(lines: [line], in: NSRange(location: 0, length: 1), parentGroup: nil))

        #expect(ObjectIdentifier(type(of: command)) == ObjectIdentifier(expected))
    }

    @Test("Only the command text is archived, and it reads back")
    func archiveRoundTrip() throws {
        let encoder = MockArchiver()
        let command = LDrawMetaCommand()
        command.commandString = "Some string"

        command.encode(with: encoder)

        #expect(encoder.data.count == 1)
        #expect(encoder.data["commandString"] as? String == "Some string")

        let decoded = try #require(LDrawMetaCommand(coder: encoder))
        #expect(decoded.commandString == "Some string")
    }

    @Test("A copy has the same text and is a separate object")
    func copyIsSeparate() throws {
        let command = LDrawMetaCommand()
        command.commandString = "Some string"

        let duplicate = try #require(command.copy() as? LDrawMetaCommand)
        #expect(duplicate.commandString == "Some string")
        #expect(duplicate !== command)

        command.commandString = "Another string"
        #expect(duplicate.commandString == "Some string")
    }

    @Test("A marker no subclass knows gives no command")
    func unknownMarkerGivesNothing() {
        let command: LDrawMetaCommand? = perform("metaCommandInstanceByMarker:scanner:", on: LDrawMetaCommand.self,
                                                 with: "string", Scanner(string: ""))

        #expect(command == nil)
    }
}
